rom 9a319e5d593fd09028b14901ba7a8110921d21a0 Mon Sep 17 00:00:00 2001
From: Julien Cristau <jcristau@debian.org>
Date: Thu, 15 Jul 2010 15:26:02 +0100
Subject: [PATCH] gio: don't assume that SOCK_CLOEXEC is supported whenever it's defined

Just because SOCK_CLOEXEC was defined at build time doesn't mean the
kernel we're running on supports it.  So if socket() fails with EINVAL,
try again without the flag.

Signed-off-by: Julien Cristau <jcristau@debian.org>
---
 gio/gsocket.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/gio/gsocket.c b/gio/gsocket.c
index e2ba788..3879202 100644
--- a/gio/gsocket.c
+++ b/gio/gsocket.c
@@ -455,9 +455,12 @@ g_socket_create_socket (GSocketFamily   family,
     }
 
 #ifdef SOCK_CLOEXEC
-  native_type |= SOCK_CLOEXEC;
+  fd = socket (family, native_type | SOCK_CLOEXEC, protocol);
+  if (fd < 0 && errno == EINVAL)
 #endif
-  fd = socket (family, native_type, protocol);
+  {
+    fd = socket (family, native_type, protocol);
+  }
 
   if (fd < 0)
     {
-- 
1.7.1

